File: frontend-next/src/types/account.ts
export type AccountType = 'demo' | 'real'

/** One trading account, with its live margin picture.
 *
 * A user may hold several of each type; `account_number` is the stable public
 * identifier and `label` is their own name for it. */
export interface TradingAccount {
  id: string
  account_number: string
  label: string | null
  currency: string
  account_type: AccountType
  is_primary: boolean
  is_active: boolean

  /** Settled cash, credits included. */
  balance: string
  /** Locked behind a pending or approved withdrawal. */
  held: string
  bonus: string
  /** Balance less outstanding bonus and reserved margin. */
  withdrawable: string
  /** Balance plus unrealised P&L. */
  equity: string

  leverage: number
  margin_used: string
  free_margin: string
  /** equity / margin_used as a percentage. Null when nothing is open — which
   * is not the same as zero, and must not be rendered as a margin problem. */
  margin_level: string | null
  margin_call: boolean
  stop_out: boolean

  unrealized_pnl: string
  realized_pnl: string
  total_pnl: string
  open_positions: number
  created_at: string
}

export interface EquityPoint {
  taken_at: string
  balance: string
  equity: string
  margin_used: string
  unrealized_pnl: string
  realized_pnl: string
}

export interface AccountCreatePayload {
  account_type: AccountType
  label?: string
  currency?: string
}

File: backend/app/schemas/account.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import AccountType


class AccountOut(BaseModel):
    """One account card: identity, money, and the margin picture."""

    id: uuid.UUID
    account_number: str
    label: str | None = None
    currency: str
    account_type: AccountType
    is_primary: bool
    is_active: bool

    balance: Decimal
    #: Locked behind a pending or approved withdrawal.
    held: Decimal = Decimal(0)
    bonus: Decimal
    #: Balance less outstanding bonus and reserved margin.
    withdrawable: Decimal
    #: Balance plus unrealised P&L.
    equity: Decimal

    leverage: int
    margin_used: Decimal
    free_margin: Decimal
    #: equity / margin_used as a percentage. None when nothing is open, which
    #: is not the same as zero.
    margin_level: Decimal | None = None
    margin_call: bool = False
    stop_out: bool = False

    unrealized_pnl: Decimal
    realized_pnl: Decimal
    total_pnl: Decimal
    open_positions: int
    created_at: datetime


class AccountCreate(BaseModel):
    account_type: AccountType = AccountType.DEMO
    label: str | None = Field(default=None, max_length=60)
    currency: str | None = Field(default=None, max_length=10)
    leverage: int = 1


class AccountUpdate(BaseModel):
    """What the account holder may change themselves. Leverage is absent on
    purpose — it is set by an admin."""

    label: str | None = Field(default=None, max_length=60)
    is_primary: bool | None = None


class AdminLeverageUpdate(BaseModel):
    leverage: int
    reason: str | None = Field(default=None, max_length=500)


class AdminBonusUpdate(BaseModel):
    """Credit or remove promotional funds."""

    amount: Decimal
    reason: str = Field(max_length=500)


class EquityPoint(BaseModel):
    taken_at: datetime
    balance: Decimal
    equity: Decimal
    margin_used: Decimal
    unrealized_pnl: Decimal
    realized_pnl: Decimal

    model_config = {"from_attributes": True}
